### PR TITLE
Add json-codec for application/grpc+json

### PR DESCRIPTION
**Pull request by zbysir**
## Motivation

A [`Codec`] that implements `application/grpc+json` via the serde library.

### CODE REVIEW

**Comment by olix0r** [comment]
This introduces dependencies that are only needed when this extra functionality is desired. Can it be feature-gated?

**Comment by zbysir** [comment]
> This introduces dependencies that are only needed when this extra functionality is desired. Can it be feature-gated?

Good idea, I will try to add ‘feature-gated’

**Comment by zbysir** [comment]
> Make json-codec be feature-gated


> This introduces dependencies that are only needed when this extra functionality is desired. Can it be feature-gated?

I did it, please review it when you have time, thank you for your help.

File: tonic/src/codec/json.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        let s = serde_json::to_string(&item).expect("Message only errors if not enough space");
+        buf.put(s.as_bytes());
+
+        Ok(())
+    }
+}
+
+pub struct JsonDecoder<U>(PhantomData<U>);
+
+impl<U: for<'a> serde::Deserialize<'a>> Decoder for JsonDecoder<U> {
```
Comment (LucioFranco) [inline tonic/src/codec/json.rs:26]: Can we use this https://docs.rs/serde/latest/serde/de/trait.DeserializeOwned.html should be the same just a bit cleaner

**Review by LucioFranco — CHANGES_REQUESTED** [review]
I would be happy to include this. But I think first I would like this PR to have an example and some tests. Then we can merge it.

**Comment by zbysir** [comment]
> I would be happy to include this. But I think first I would like this PR to have an example and some tests. Then we can merge it.

Okay, I will add some tests and example later.

**Comment by zbysir** [comment]
Hello?

**Review by LucioFranco — CHANGES_REQUESTED** [review]
Sorry for the delay. I think I would like to merge this on the condition that we add the change for build and add an example + tests (I think you added tests so that is good!).

**Pull request closed without merging**